mypy/checkexpr.py

```diff
@@ -5082,15 +5087,27 @@ def fast_container_type(
                 # fallback to slow path
                 self.resolved_type[e] = NoneType()
                 return None
-            values.append(self.accept(item))
-        vt = join.join_type_list(values)
-        if not allow_fast_container_literal(vt):
+
+            typ = self.accept(item)
+            if typ not in values:
```

Comment:
This will be quadratic for long lists. Maybe it would make sense to have a set here (and similar for dicts) in addition to the list?